File: 30_day_challenge_2020_December/80_remove_duplicates_from_sorted_array_ii.py
################################################################
# Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
# Date created      : 20201211
# Problem link      : https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
################################################################

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        i, n, c = 0, None, 0 # i: index of pos for new value, n: the last num, c: times last num encountered
        for j in range(len(nums)):
            if n != nums[j]:
                n, c = nums[j], 0
            c += 1
            if c <= 2:
                nums[i] = nums[j]
                i += 1
        return i
        
#     # @StefanPochmann
#     def removeDuplicates(self, nums):
#         i = 0
#         for n in nums:
#             if i < 2 or n > nums[i-2]:
#                 nums[i] = n
#                 i += 1
#         return i

    def removeDuplicates(self, nums):
        slow, fast = 2, 2
        while fast < len(nums):
            if nums[slow - 2] != nums[fast]:
                nums[slow] = nums[fast]
                slow += 1
            fast += 1
        return slow

File: 30_day_challenge_2020_December/80_remove_duplicates_from_sorted_array_ii.cpp
/****************************************************************
* Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
* Date created      : 20201211
* Problem link      : https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
****************************************************************/

class Solution {
public:
    // int removeDuplicates(vector<int>& nums) {
    //     if (nums.size() < 2)
    //         return nums.size();
    //     int s = 2, f = 2; // slow, fast pointer
    //     while (f < nums.size()){
    //         if (nums[s-2] < nums[f]) { // if 2nd last number is different than new one
    //             nums[s] = nums[f];
    //             s += 1;
    //         }
    //         f += 1;
    //     }
    //     return s;
    // }
    
    // @StefanPochmann
    int removeDuplicates(vector<int>& nums) {
        int i = 0;
        for (int n : nums)
            if (i < 2 || n > nums[i-2])
                nums[i++] = n;
        return i;
    }
};